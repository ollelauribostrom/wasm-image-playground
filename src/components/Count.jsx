import React from 'react';
import { connect } from '../store/AppStore';

type CountProps = {|
  count: number
|};

function Count({ count }: CountProps) {
  const times = count === 1 ? 'TIME' : 'TIMES';
  const message = count === 0 ? 'CLICK THE REACTOR' : `CLICKED ${count} ${times}`;
  return <div className="reactor-count">{message}</div>;
}

export default connect(state => state)(Count);
