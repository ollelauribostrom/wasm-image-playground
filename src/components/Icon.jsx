import React from 'react';
import icons from '../assets';

function Icon({ name, alt = "icon", size, style, className, ...props }) {
  return (
    <img 
      src={icons[name]} 
      alt={alt}
      style={style}
      className={`icon icon-${size} ${className}`}
      {...props}
    />)
}

export default Icon;