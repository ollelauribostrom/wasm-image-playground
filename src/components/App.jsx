import React from 'react';
import { Provider } from '../store/AppStore';
import ErrorBoundary from './ErrorBoundary';
import Reactor from './Reactor';
import Count from './Count';
import '../styles/main.css';

export default function App() {
  return (
    <Provider>
      <ErrorBoundary>
        <Reactor />
        <Count />
      </ErrorBoundary>
    </Provider>
  );
}
