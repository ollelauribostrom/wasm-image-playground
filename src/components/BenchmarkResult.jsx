import React from 'react';
import Icon from './Icon';

function BenchmarkResult({ task, fastest, percent, avgDiff }) {
  const diff = avgDiff !== undefined ? `(${avgDiff}% difference in output pixels)` : '';
  const info = `${task}: ${fastest} was ${percent.total}% faster ${diff}`;
  return (
    <div className="benchmark-result" key={task}>
      {<Icon name="results" size="s" />}
      <span className="benchmark-result-info">{info}</span>
    </div>
  )
}

export default BenchmarkResult;
