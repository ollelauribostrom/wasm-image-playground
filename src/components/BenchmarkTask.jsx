import React from 'react';
import Icon from './Icon';
import Spinner from './Spinner';

function getStatusIcon(status) {
  if (status === 'running') {
    return (
      <div className="benchmark-task-status-icon">
        <Spinner visible={true} color="#A599FF" />
      </div>
    );
  }
  if (status === 'done') {
    return <Icon name="check" size="s" className="benchmark-task-status-icon" />;
  }
  if (status === 'error') {
    return <Icon name="cross" size="s" className="benchmark-task-status-icon" />;
  }
}

function BenchmarkTask({ id, status, info, time, average, type, iconSize = 's' }) {
  return (
    <div className="benchmark-task" key={id}>
      {<Icon name={type} size={iconSize} />}
      <span className="benchmark-info">{info}</span>
      {time ? <span className="benchmark-time">Total: {time}ms</span> : null}
      {average ? <span className="benchmark-average">Average: {average}ms</span> : null }
      {getStatusIcon(status)}
    </div>
  )
}

export default BenchmarkTask;
