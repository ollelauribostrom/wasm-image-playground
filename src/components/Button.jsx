import React from 'react';
import { Icon } from './Icon';
import { ToggleHandle } from './ToggleHandle';

type ButtonProps = {
  icon?: Icon,
  label?: string,
  title?: string,
  isDisabled?: boolean,
  onClick?: () => void,
  onToggle?: () => void,
  customClassName?: string,
  href?: string,
  target?: string,
  rel?: string
};

export function Button({
  icon,
  label,
  title,
  isDisabled,
  onClick,
  onToggle,
  customClassName,
  href,
  target,
  rel
}: ButtonProps) {
  let className = `button ${customClassName}`;
  if (isDisabled) {
    className += 'button--disabled';
  }
  if (href && !isDisabled) {
    return (
      <a className={className} href={href} target={target} rel={rel} title={title}>
        {label}
        {icon}
      </a>
    );
  }
  return (
    <div
      onKeyPress={e => {
        const key = e.key || e.keyCode;
        if (key === 'Enter' || key === 13) {
          if (!isDisabled) {
            onClick();
          }
        }
      }}
      onClick={e => !isDisabled && onClick(e)}
      className={`button ${customClassName}`}
      type="button"
      title={title}
      role="button"
      tabIndex={0}
    >
      {label ? <span>{label}</span> : null}
      {icon}
      {onToggle ? (
        <ToggleHandle
          onToggle={e => {
            e.stopPropagation();
            onToggle();
          }}
        />
      ) : null}
    </div>
  );
}

Button.defaultProps = {
  icon: null,
  label: null,
  title: '',
  isDisabled: false,
  onClick: () => {},
  onToggle: null,
  customClassName: '',
  href: null,
  target: null,
  rel: null
};
