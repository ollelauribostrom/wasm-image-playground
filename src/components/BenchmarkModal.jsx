import React from 'react';
import Modal from 'react-modal';
import Icon from './Icon';
import BenchmarkTask from './BenchmarkTask';
import BenchmarkResult from './BenchmarkResult';
import Label from './Label';

const modalStyle = {
  content: {
    background: "transparent",
    border: 0
  },
  overlay: {
    background: "rgba(0, 0, 0, 0.9)",
    border: 0,
    zIndex: 2,            
  }
};

function BenchmarkModal({ isOpen, title, running, results, tasks, onStart, onRestart, onClose }) {
  return (
    <Modal isOpen={isOpen} style={modalStyle} >
      <div className="benchmark-content">
        <h1 className="benchmark-heading">Benchmark: {title}</h1>
        <Icon 
          name="restore"
          size="l"
          onClick={onRestart}
          className={`benchmark-reset-icon ${!running && results.length ? '' : 'hidden'}`}
        />
        <Icon 
          name="close"
          size="l"
          onClick={onClose}
          className="benchmark-close-icon"
        />
        <Label 
          text="Run Benchmark"
          size="large"
          onClick={onStart}
          className={`benchmark-button ${!running && !results.length ? '' : 'hidden'}`} 
        />
        {tasks.map(task => <BenchmarkTask {...task} />)}
        <div className={`benchmark-results ${results.length ? '' : 'hidden'}`}>
          {results.map(result => <BenchmarkResult {...result} />)}
        </div>
      </div>
    </Modal>
  )
}

export default BenchmarkModal;