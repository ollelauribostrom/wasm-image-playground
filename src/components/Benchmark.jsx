import React, { Component } from 'react';
import { imageConverters } from 'imutils';
import BenchmarkModal from './BenchmarkModal';
import ImageService from '../services/ImageService';
import benchmarkImages from '../../benchmark/images.json'

class Benchmark extends Component {
  state = {
    taskIndex: 0,
    tasks: [],
    taskResults: {},
    results: [],
    error: null,
    running: false,
    showResult: false
  }

  componentWillMount() {
    ImageService.on('benchmarkUpdate', this.onUpdate);
    ImageService.on('benchmarkError', this.onError);
    ImageService.on('benchmarkComplete', this.onComplete);
  }

  componentWillUnmount() {
    ImageService.removeListener('benchmarkUpdate', this.onUpdate);
    ImageService.removeListener('benchmarkError', this.onError);
    ImageService.removeListener('benchmarkComplete', this.onComplete);
  }

  restart = async () => {
    if (!this.state.running && this.state.results) {
      await this.setState({ tasks: [], results: [], error: null, taskIndex: 0, taskResults: {} });
      this.start();
    }
  }

  start = () => {
    this.props.onStart();
    this.setState({ running: true });
    this.setup();
  }

  stop = () => {
    this.props.onStop();
  }

  setup = async () => {
    try {
      this.addOrUpdateTask({
        id: 'SETUP',
        info: 'Loading benchmark images',
        status: 'running',
        type: 'setup'
      });
      const images = await Promise.all(benchmarkImages.files.map(async image => {
        const res = await fetch(require(`../../benchmark/${image.path}`));
        const blob = await res.blob();
        const data = await imageConverters.toUint8ClampedArray(URL.createObjectURL(blob));
        return { data, containsFace: image.containsFace }
      }));
      await this.setState({ images });
      this.addOrUpdateTask({
        id: 'SETUP',
        info: `Loaded ${images.length} benchmark images`,
        status: 'done',
      });
      this.startTask(this.props.tasks[0]);
    } catch (error) {
      this.addOrUpdateTask({
        id: 'SETUP',
        info: 'Error when loading benchmark images',
        status: 'error',
      });
    }
  }

  startTask = task => {
    ImageService.postMessage({
      task,
      action: this.props.benchmarkType,
      images: this.state.images,
      results: task === 'getResults' ? this.state.taskResults : null
    });
  }

  addOrUpdateTask = task => {
    const tasks = [...this.state.tasks];
    const taskIndex = tasks.findIndex(t => t.id === task.id);
    if (taskIndex > -1) {
      tasks[taskIndex] = Object.assign(tasks[taskIndex], task);
    } else {
      tasks.push(task)
    }
    this.setState({ tasks });

    if (task.status === 'done' && task.id !== 'SETUP') {
      const taskResults = Object.assign({}, this.state.taskResults);
      const currentTask = this.props.tasks[this.state.taskIndex];
      const nextTask = this.props.tasks[this.state.taskIndex + 1];
      taskResults[currentTask] = task.result;
      this.setState({ taskResults, taskIndex: this.state.taskIndex + 1 });
      if (nextTask) {
        this.startTask(nextTask);
      }
    }
  }

  onUpdate = ({ data }) => {
    this.addOrUpdateTask(data.task);
  }
  
  onError = ({ data }) => {
    this.setState({
      error: data.error,
      running: false
    });
    this.props.onStop();
  }

  onComplete = ({ data }) => {
    this.setState({ running: false, results: data.results });
    this.props.onStop();
  }

  onClose = () => {
    this.setState({
      isOpen: false
    });
  }

  render() {
    return (
      <BenchmarkModal
        isOpen={this.props.isOpen}
        running={this.state.running}
        tasks={this.state.tasks}
        results={this.state.results}
        title={this.props.title}
        onStart={this.start}
        onRestart={this.restart}
        onClose={this.props.onClose}
      />
    )
  }
}

export default Benchmark;
