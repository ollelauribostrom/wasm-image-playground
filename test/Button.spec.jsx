import React from 'react';
import { shallow, mount } from 'enzyme';
import { Button } from '../src/components/Button';
import { Icon } from '../src/components/Icon';

describe('Tests for Button', () => {
  it('renders correctly', () => {
    const wrapper = shallow(
      <Button customClassName="testClass" title="testTitle" icon={<Icon name="image" />} />
    );
    expect(wrapper).toMatchSnapshot();
  });
  it('renders correctly when passing the isDisabled prop', () => {
    const wrapper = shallow(
      <Button customClassName="testClass" icon={<Icon name="image" />} isDisabled />
    );
    expect(wrapper).toMatchSnapshot();
  });
  it('renders correctly when passing the href prop', () => {
    const wrapper = shallow(
      <Button
        label="label"
        href="testHref"
        target="testTarget"
        rel="testRel"
        icon={<Icon name="image" />}
      />
    );
    expect(wrapper).toMatchSnapshot();
  });
  it('renders correctly when passing the onToggle prop', () => {
    const onToggle = jest.fn();
    const wrapper = shallow(
      <Button
        className="testClass"
        title="testTitle"
        icon={<Icon name="image" />}
        onToggle={onToggle}
      />
    );
    expect(wrapper).toMatchSnapshot();
  });
  it('invokes onClick when clicked', () => {
    const onClick = jest.fn();
    const wrapper = shallow(<Button onClick={onClick} />);
    wrapper.simulate('click');
    expect(onClick).toHaveBeenCalledTimes(1);
  });
  it('invokes onClick on enter keypress', () => {
    const onClick = jest.fn();
    const wrapper = shallow(<Button onClick={onClick} />);
    wrapper.simulate('keypress', { keyCode: 13 });
    wrapper.simulate('keypress', { keyCode: 12 });
    wrapper.simulate('keypress', { key: 'Enter' });
    expect(onClick).toHaveBeenCalledTimes(2);
  });
  it('invokes onToggle when clicking the toggle handle', () => {
    const onClick = jest.fn();
    const onToggle = jest.fn();
    const wrapper = mount(<Button onClick={onClick} onToggle={onToggle} />);
    wrapper.find('ToggleHandle').simulate('click');
    expect(onClick).not.toHaveBeenCalled();
    expect(onToggle).toHaveBeenCalledTimes(1);
    wrapper.unmount();
  });
  it('does not invoke onClick if the button is disabled', () => {
    const onClick = jest.fn();
    const wrapper = shallow(<Button onClick={onClick} isDisabled />);
    wrapper.simulate('click');
    wrapper.simulate('keypress', { key: 'Enter' });
    expect(onClick).not.toHaveBeenCalled();
  });
});
