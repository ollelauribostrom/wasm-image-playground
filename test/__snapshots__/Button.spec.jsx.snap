// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests for Button renders correctly 1`] = `
<div
  className="button testClass"
  onClick={[Function]}
  onKeyPress={[Function]}
  role="button"
  tabIndex={0}
  title="testTitle"
  type="button"
>
  <Icon
    name="image"
  />
</div>
`;

exports[`Tests for Button renders correctly when passing the href prop 1`] = `
<a
  className="button "
  href="testHref"
  rel="testRel"
  target="testTarget"
  title=""
>
  label
  <Icon
    name="image"
  />
</a>
`;

exports[`Tests for Button renders correctly when passing the isDisabled prop 1`] = `
<div
  className="button testClass"
  onClick={[Function]}
  onKeyPress={[Function]}
  role="button"
  tabIndex={0}
  title=""
  type="button"
>
  <Icon
    name="image"
  />
</div>
`;

exports[`Tests for Button renders correctly when passing the onToggle prop 1`] = `
<div
  className="button "
  onClick={[Function]}
  onKeyPress={[Function]}
  role="button"
  tabIndex={0}
  title="testTitle"
  type="button"
>
  <Icon
    name="image"
  />
  <ToggleHandle
    onToggle={[Function]}
  />
</div>
`;
